### ISSUE TITLE
SweepStep is missing opportunities to sweep some metadata

### ISSUE THREAD

**Issue by spouliot** [mono/linker#1188]
### ModuleReferences - SUPPORTED

Only when linking should we remove module references, if we (re)save the assembly then
the entry points (for p/invokes) will be required later
reference: https://bugzilla.xamarin.com/show_bug.cgi?id=35372

This is also **critical** for AppStore submissions since it will keep references (full paths) to all native (bound) frameworks (even if all the related API are removed) and some of them requires special entitlements (or extra setup) to be accepted.

Existing implementation: [xamarin-macios](https://github.com/xamarin/xamarin-macios/blob/master/tools/linker/MobileSweepStep.cs#L56)

### ExportedTypes - SUPPORTED

if we save (only or by linking) then unmarked exports (e.g. forwarders) must be cleaned
or they can point to nothing which will break later (e.g. when re-loading for stripping IL)
reference: https://bugzilla.xamarin.com/show_bug.cgi?id=36577

Existing implementation: [xamarin-macios](https://github.com/xamarin/xamarin-macios/blob/3549ccf3dda90989a20f5cf28125623161d093c7/tools/linker/MobileSweepStep.cs#L51)

### InternalsVisibleToAttribute - SUPPORTED

References to removed assemblies still exists inside `[InternalsVisibleTo]` attributes.

Existing implementation: [xamarin-macios](https://github.com/xamarin/xamarin-macios/blob/3549ccf3dda90989a20f5cf28125623161d093c7/tools/linker/MonoTouch.Tuner/MonoTouchSweepStep.cs#L26)

**Comment by spouliot** [comment]
Looks like `ExportedTypes` are already handled.

**Comment by marek-safar** [comment]
> ModuleReferences

Moved the logic in #1193

> ExportedTypes

It should be already done

> InternalsVisibleToAttribute

I don't think we have to keep InternalsVisibleToAttribute at all. It's compiler attribute which could be always removed. The simplest way will be to use #1162 once it lands

**Comment by MichalStrehovsky** [comment]
> I don't think we have to keep InternalsVisibleToAttribute at all. It's compiler attribute which could be always removed. The simplest way will be to use #1162 once it lands

This is used to enforce visibility/accessibility checks within the runtime. Dropping it would fail the access checks and throw exceptions at runtime. https://github.com/dotnet/runtime/blob/57ad64d4de6cd6c6de4ffb7fb6176d46b0da9d87/src/coreclr/src/vm/assembly.cpp#L2381

**Comment by marek-safar** [comment]
> Dropping it would fail the access checks and throw exceptions at runtime

Does reflection API use that or something else?

**Comment by MichalStrehovsky** [comment]
> Does reflection API use that or something else?

Don't know about reflection, but e.g. if the IL has e.g. a `ldfld` instruction of field that is not accessible from the method, a `FieldAccessException` is going to be thrown, per ECMA-335.

**Comment by marek-safar** [comment]
All improvements are available in master

**Comment by spouliot** [comment]
Re-opening as the IVT part was reverted.

**Comment by marek-safar** [comment]
All the opportunities are supported in .net6
